### PR TITLE
Added middleware endpoint

### PR DESCRIPTION
**Pull request by AlfonsoUceda**
lotus/controller#72 Create a middleware endpoint for an action with one or several middlewares.

This is necessary for creating new PR in lotus/controller, fixing the rack.rb class.

Example:

```ruby
require 'lotus/router'
require 'lotus/controller'

class MyMiddleware
  def initialize(app)
   @app = app
  end

  def call(env)
    code, headers, body = @app.call(env)
    [code, headers.merge('X-Middleware' => 'CALLED'), body]
  end
end

module Home
  class Index
    include Lotus::Action
    use MyMiddleware

    def call(params)
      self.body = 'foo'
    end
  end
end

app = Lotus::Router.new do
  get '/', to: 'home#index'
end

Rack::Server.start app: app, Port: 2300
```
I haven't added new tests, ClassEndpoint isn't tested, how is it possible to test this MiddlewareEndpoint?

### CODE REVIEW

**Comment by jodosha** [comment]
@AlfonsoUceda Thanks for this PR.

Please have a look at https://github.com/lotus/router/blob/master/test/routing/endpoint_resolver_test.rb

**Comment by AlfonsoUceda** [comment]
@jodosha what do you think about these tests?

File: lib/lotus/routing/endpoint_resolver.rb [hunk 1]
```diff
@@ -197,7 +198,7 @@ def classify(string)
       private
       def resolve_callable(callable)
         if callable.respond_to?(:call)
-          @endpoint_class.new(callable)
+          create_endpoint(callable, @endpoint_class)
```
Comment (AlfonsoUceda) [inline lib/lotus/routing/endpoint_resolver.rb]: @jodosha  do you think create_endpoint here is necessary? because always inside create_endpoint is checking again the respond_to and it always goes through `MiddlewareEndpoint`

Comment (jodosha) [inline lib/lotus/routing/endpoint_resolver.rb]: @AlfonsoUceda We can review it later, when the PR for Controller will be in place ;)

**Comment by jodosha** [comment]
Things are simpler than I expected :wink:

There are two symptoms that led me to this finding:
  1. Code duplication: #create_endpoint and #resolve_callable are doing the same thing.
  2. We are testing implementation instead of behaviors: https://github.com/lotus/router/pull/36/files#diff-a3ea2e007d31d69e1fd367fc500263b2R65

I found that `MiddlewareEndpoint` isn't needed at all. `Endpoint` is enough.
Because it inheriths from `SimpleDelegator`, so all the missing methods are forwarded to the wrapped object.

```ruby
def call(env)
  __getobj__.call(env)
end
```

Can be replaced with

```ruby
def call(env)
  super
end
```

That means, we can remove it entirely.
We just be smart enough to pass the right object to `Endpoint`.

Now, the tests that we put in place aren't covering **what** we want to achieve, but **how** we want to achieve it.

In other words, what we want to do is expressed by high level tests, usually integration tests.
Where we simulate a Lotus action with an inner middleware, and a router that forwards HTTP requests to it.

That's basically our feature (the **"what"**).

Testing if a returned object is of a given type is useful in certain scenarios, but it's always focused about the details (the **"how"**).

To recap, we probably need another integration test here. :smile:

We still can use `Endpoint`
And probably we should **only** do this:

```ruby
def resolve_matchable(matchable)
  if matchable.respond_to?(:match)
    constantize(
      resolve_callable(resolve_action(matchable)) || classify(matchable)
    )
  end
end
```

I learned a great lesson. @AlfonsoUceda thanks for making this possible. :+1:

File: lib/lotus/routing/endpoint.rb [hunk 2]
```diff
@@ -44,6 +44,11 @@ def inspect
           "#<#{ __getobj__.class }>"
         end
       end
+
+      # @since x.x.x
+      def call(env)
```
Comment (jodosha) [inline lib/lotus/routing/endpoint.rb]: This isn't necessary, because `Endpoint` inherits from `SimpleDelegator` this behavior **already happens**. ;)

File: test/integration/middleware_integration_test.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    response.body.must_equal 'Hello from Web::Controllers::Home::Index'
+    response.status.must_equal 200
+    response.headers.fetch('X-Middleware').must_equal 'CALLED'
+  end
+
+  it 'action without middleware' do
+    response = @app.get('/dashboard')
+    response.body.must_equal 'Hello from Web::Controllers::Dashboard::Index'
+    response.status.must_equal 200
+    -> { response.headers.fetch('X-Middleware') }.must_raise KeyError
```
Comment (jodosha) [inline test/integration/middleware_integration_test.rb]: There is no need to raise an exception here, because we already know that the key is missing.
What if we change it with?

```ruby
response.headers['X-Middleware'].must_be_nil
```

:smile_cat:

**Comment by jodosha** [comment]
This looks good :+1: Thank you!
